src/recovery/mod.rs

```diff
@@ -0,0 +1,282 @@
+        (Self { db_file_name }, RecoveryConfig {})
+    }
+
+    /// Pickle as a tuple.
+    fn __getstate__(&self) -> (&str, &str) {
+        ("SqliteRecoveryConfig", &self.db_file_name)
+    }
+
+    /// Egregious hack because pickling assumes the type has "empty"
+    /// mutable objects.
+    ///
+    /// Pickle always calls `__new__(*__getnewargs__())` but notice we
+    /// don't have access to the pickled `db_file_name` yet, so we
+    /// have to pass in some dummy string value that will be
+    /// overwritten by `__setstate__()` shortly.
+    fn __getnewargs__(&self) -> (&str,) {
+        ("UNINIT_PICKLED_STRING",)
+    }
+
+    /// Unpickle from tuple of arguments.
+    fn __setstate__(&mut self, state: &PyAny) -> PyResult<()> {
+        if let Ok(("SqliteRecoveryConfig", db_file_name)) = state.extract() {
+            self.db_file_name = db_file_name;
+            Ok(())
+        } else {
+            Err(PyValueError::new_err(format!(
+                "bad pickle contents for SqliteRecoveryConfig: {state:?}"
+            )))
+        }
+    }
+}
+
+/// Rust-side trait which represents actions the dataflow execution
+/// will need to delegate to a recovery store.
+///
+/// We'll implement this for each kind of recovery store we need to
+/// talk to. This is not exposed Python-side.
+pub(crate) trait RecoveryStore {
+    // TODO: I wanted to make a trait RecoveryStore statically typed, but I
+    // think that'll require GAT / generic associated types.
+
+    /// Build recovery interface for this step.
+    fn for_step(&self, step_id: &str) -> Box<dyn StepRecovery<u64, TdPyAny, TdPyAny>>;
+
+    // TODO: We can implement state compaction / GC once we know
+    // epochs are output by every capture() operator.
+    //fn compact(&self, done_epoch: u64);
+}
+
+/// Per-operator interface for state recovery.
+///
+/// Each [`RecoveryStore`] will need to implement this with the
+/// specific queryies or inserts needed.
+pub(crate) trait StepRecovery<T, K, D> {
+    fn recover_last(&self, current_epoch: &T, key: &K) -> Option<D>;
+
+    fn save_complete(&self, completed_epoch: &T, key: &K, state: &Option<D>) -> ();
+}
+
+/// A recovery store which does nothing.
+///
+/// Saves are dropped and all recoveries result in "not found".
+pub(crate) struct NoOpRecoveryStore;
+
+impl NoOpRecoveryStore {
+    pub(crate) fn new() -> Self {
+        NoOpRecoveryStore {}
+    }
+}
+
+impl RecoveryStore for NoOpRecoveryStore {
+    fn for_step(&self, step_id: &str) -> Box<dyn StepRecovery<u64, TdPyAny, TdPyAny>> {
+        Box::new(NoOpStepRecovery {
+            step_id: step_id.to_string(),
+        })
+    }
+}
+
+struct NoOpStepRecovery {
+    step_id: String,
+}
+
+impl<T: Debug, K: Debug, D: Debug> StepRecovery<T, K, D> for NoOpStepRecovery {
+    fn recover_last(&self, current_epoch: &T, key: &K) -> Option<D> {
+        debug!(
+            "noop state fake-recovered step_id={} key={key:?}@epoch={current_epoch:?}",
```

Comment:
When would we want to see this? If we did like a dry run? I got stuck on the "fake-recovered" bit

Reply:
It's "fake" in that the no-op store will never be able to recover state. It always returns "not found". I wanted to log that a query to the recovery store occurred, though. This is mostly useful for debugging Bytewax and state recovery itself. Let me use some clearer words here. Eventually we should namespace our logging so someone can tune the spew.

This touches on a good point in that I'm sort of overloading the concept of recovery here. There's nothing stopping one from "recovering" a dataflow execution using this no-op store and there being no errors, but you'd get different output because all of the state was reset to "new key". Although no one is stopping you from re-running your dataflow over different data or starting at a different point in the input and seeing different output. I guess it's important to realize that output is a function of:

1. Input values
2. Input order
3. Recovery store state

All at once. You can't really detect that one is "wrong".